### PR TITLE
feat: support async app template

### PR DESCRIPTION
**Pull request by marvinhagemeister**
This PR adds support for async `_app` templates. It was the last thing which didn't support that. Now all routes, layouts or the app template all behave the same internally. In doing so I noticed that we can make the rendering function a lot simpler and make a bunch of error checks redundant.

### CODE REVIEW

**Comment by marvinhagemeister** [comment]
FYI: I've simplified a bit the async route component TS types too, since they all have the same args, except for a route not having `Component`

File: tests/fixture_layouts/routes/async/_layout.tsx [hunk 1]
```diff
@@ -1,15 +1,14 @@
-import { LayoutProps, RouteContext } from "$fresh/server.ts";
+import { LayoutContext } from "$fresh/server.ts";
 
 export default async function AsyncLayout(
   req: Request,
-  ctx: RouteContext,
-  { Component }: LayoutProps,
+  ctx: LayoutContext,
 ) {
   await new Promise((r) => setTimeout(r, 10));
```
Comment (iuioiua) [inline tests/fixture_layouts/routes/async/_layout.tsx]:
```suggestion
  await delay(10);
```
Nit. Just for consistency.

File: src/server/types.ts [hunk 2]
```diff
@@ -162,7 +170,7 @@ export interface Route<Data = any> {
   pattern: string;
   url: string;
   name: string;
-  component?: PageComponent<Data>;
+  component?: PageComponent<Data> | AsyncLayout | AsyncRoute;
```
Comment (iuioiua) [inline src/server/types.ts]:
```suggestion
  component?: PageComponent<Data> | AsyncLayout<Data> | AsyncRoute<Data>;
```

**Review by iuioiua — COMMENTED** [review]
I don't entirely understand the implementation but spotted a couple nits 🙂

File: tests/fixture_layouts/routes/async/_layout.tsx (hunk 1 above)
Comment (marvinhagemeister) [reply tests/fixture_layouts/routes/async/_layout.tsx]: At some point I'll remember that we have this in our standard library 👀

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**